mempool/src/pool.rs

```diff
@@ -0,0 +1,2787 @@
+        //
+        // - Bitcoin Core does not relay transactions smaller than 82 bytes (see
+        // MIN_STANDARD_TX_NONWITNESS_SIZE in Bitcoin Core's policy.h)
+        //
+        // - Checking that coinbase inputs have matured
+        //
+        // - Checking the signature operations cost (Bitcoin Core: `GetTransactionSigOpCost`)
+        //
+        // - We have yet to understand and implement this comment pertaining to chain limits and
+        // calculation of in-mempool ancestors:
+        // https://github.com/bitcoin/bitcoin/blob/7c08d81e119570792648fe95bbacddbb1d5f9ae2/src/validation.cpp#L829
+        //
+        // - Bitcoin Core's `EntriesAndTxidsDisjoint` check
+        //
+        // - Bitcoin Core's `PolicyScriptChecks`
+        //
+        // - Bitcoin Core's `ConsensusScriptChecks`
+
+        // Deviations from Bitcoin Core
+        //
+        // - In our FeeRate calculations, we use the `encoded_size`of a transaction. In contrast,
+        // see Bitcoin Core's `CTxMemPoolEntry::GetTxSize()`, which takes into account sigops cost
+        // and witness data. This deviation is not intentional, but rather the result of wanting to
+        // get a basic implementation working. TODO: weigh what notion of size/weight we wish to
+        // use and whether to follow Bitcoin Core in this regard
+        //
+        // We don't have a notion of MAX_MONEY like Bitcoin Core does, so we also don't have a
+        // `MoneyRange` check like the one found in Bitcoin Core's `CheckTransaction`
+
+        if tx.get_inputs().is_empty() {
+            return Err(TxValidationError::NoInputs);
+        }
+
+        if tx.get_outputs().is_empty() {
+            return Err(TxValidationError::NoOutputs);
+        }
+
+        if tx.is_coinbase() {
+            return Err(TxValidationError::LooseCoinbase);
+        }
+
+        let outpoints = tx.get_inputs().iter().map(|input| input.get_outpoint()).cloned();
+
+        if has_duplicate_entry(outpoints) {
+            return Err(TxValidationError::DuplicateInputs);
+        }
+
+        if tx.encoded_size() > MAX_BLOCK_SIZE_BYTES {
+            return Err(TxValidationError::ExceedsMaxBlockSize);
+        }
+
+        if self.contains_transaction(&tx.get_id()) {
+            return Err(TxValidationError::TransactionAlreadyInMempool);
+        }
+
+        let conflicts = self.rbf_checks(tx)?;
+
+        self.verify_inputs_available(tx)?;
+
+        self.pays_minimum_relay_fees(tx)?;
+
+        self.pays_minimum_mempool_fee(tx)?;
+
+        Ok(conflicts)
+    }
+
+    fn pays_minimum_mempool_fee(&self, tx: &Transaction) -> Result<(), TxValidationError> {
+        let tx_fee = self.try_get_fee(tx)?;
+        let minimum_fee = self.get_update_minimum_mempool_fee(tx);
+        (tx_fee >= minimum_fee)
+            .then(|| ())
+            .ok_or(TxValidationError::RollingFeeThresholdNotMet {
+                minimum_fee,
+                tx_fee,
+            })
```

Comment:
Our `ensure` calls `?`. Here I just want to compute the result and and call `?` on the result of a function call with an informative name. So not sure how our ensure can fit in here....

Reply:
Yea it would be something like:
```rust
ensure!(
    tx_fee >= minimum_fee,
    TxValidationError::RollingFeeThresholdNotMet { minimum_fee, tx_fee },
);
Ok(())
```
Maybe not a big win